File: DP/LeetCode_44_WildcardMatching.java
public class LeetCode_44_WildcardMatching {
    // Boolean[][] dp;

    // public boolean solve(int i, int j, String s, String p){

    //     if(j == p.length())
    //         return i == s.length();

    //     if(dp[i][j] != null)
    //         return dp[i][j];

    //     boolean match = (i < s.length() &&
    //             (s.charAt(i) == p.charAt(j) || p.charAt(j) == '?'));

    //     boolean ans;

    //     if(p.charAt(j) == '*'){
    //         ans = solve(i, j + 1, s, p) ||   // skip *
    //               (i < s.length() && solve(i + 1, j, s, p)); // use *
    //     } 
    //     else {
    //         ans = match && solve(i + 1, j + 1, s, p);
    //     }

    //     return dp[i][j] = ans;
    // }

    // public boolean isMatch(String s, String p) {
    //     dp = new Boolean[s.length() + 1][p.length() + 1];
    //     return solve(0, 0, s, p);
    // }
    public boolean isMatch(String s, String p) {
        int n = s.length();
        int m = p.length();

        boolean[][] dp = new boolean[n + 1][m + 1];

        dp[0][0] = true;

        // handle pattern starting with *
        for(int j = 1; j <= m; j++){
            if(p.charAt(j - 1) == '*'){
                dp[0][j] = dp[0][j - 1];
            }
        }

        for(int i = 1; i <= n; i++){
            for(int j = 1; j <= m; j++){

                if(p.charAt(j - 1) == s.charAt(i - 1) || p.charAt(j - 1) == '?'){
                    dp[i][j] = dp[i - 1][j - 1];
                }

                else if(p.charAt(j - 1) == '*'){
                    dp[i][j] = dp[i][j - 1] || dp[i - 1][j];
                }

                else{
                    dp[i][j] = false;
                }
            }
        }

        return dp[n][m];
    }
}

File: DP/LeetCode_44_WildcardMatching.py
class Solution:
    # def isMatch(self, s: str, p: str) -> bool:
    #     # 🔥 STEP 1: remove consecutive '*'
    #     new_p = []
    #     for ch in p:
    #         if not (new_p and new_p[-1] == '*' and ch == '*'):
    #             new_p.append(ch)
    #     p = "".join(new_p)

    #     dp = {}

    #     def solve(i, j):
    #         if (i, j) in dp:
    #             return dp[(i, j)]

    #         if j == len(p):
    #             return i == len(s)

    #         if p[j] == '*':
    #             ans = (solve(i, j + 1) or
    #                    (i < len(s) and solve(i + 1, j)))
    #         else:
    #             match = (i < len(s) and (s[i] == p[j] or p[j] == '?'))
    #             ans = match and solve(i + 1, j + 1)

    #         dp[(i, j)] = ans
    #         return ans

    #     return solve(0, 0)

    def isMatch(self, s: str, p: str) -> bool:
        n, m = len(s), len(p)

        # 🔥 compress pattern (important)
        new_p = []
        for ch in p:
            if not (new_p and new_p[-1] == '*' and ch == '*'):
                new_p.append(ch)
        p = "".join(new_p)
        m = len(p)

        # dp table
        dp = [[False]*(m+1) for _ in range(n+1)]

        # base case
        dp[0][0] = True

        # pattern like *, **, ***
        for j in range(1, m+1):
            if p[j-1] == '*':
                dp[0][j] = dp[0][j-1]

        # fill table
        for i in range(1, n+1):
            for j in range(1, m+1):

                if p[j-1] == '*':
                    dp[i][j] = dp[i][j-1] or dp[i-1][j]

                elif p[j-1] == '?' or s[i-1] == p[j-1]:
                    dp[i][j] = dp[i-1][j-1]

                else:
                    dp[i][j] = False

        return dp[n][m]